File: examples/scala/ethereum-client/src/main/scala/Web3Client.scala
package com.web3.client

import scala.concurrent.{Future, ExecutionContext}
import scala.util.{Try, Success, Failure}
import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.net.http.HttpRequest.BodyPublishers
import java.net.http.HttpResponse.BodyHandlers
import spray.json._
import DefaultJsonProtocol._

/**
 * Functional Web3 client for Ethereum using Scala
 * Demonstrates type-safe blockchain interactions with functional programming
 */
object Web3Client {

  case class RpcRequest(
    jsonrpc: String = "2.0",
    method: String,
    params: List[Any],
    id: Int
  )

  case class RpcResponse[T](
    jsonrpc: String,
    result: Option[T],
    error: Option[JsValue],
    id: Int
  )

  class EthereumClient(rpcUrl: String = "https://eth.llamarpc.com") {
    private var requestId = 0
    private val client = HttpClient.newHttpClient()

    implicit val rpcRequestFormat = jsonFormat4(RpcRequest)

    /**
     * Make RPC call to Ethereum node
     */
    private def rpcCall[T: JsonReader](method: String, params: List[Any] = List.empty): Try[T] = Try {
      requestId += 1

      val request = RpcRequest(
        method = method,
        params = params,
        id = requestId
      )

      val httpRequest = HttpRequest.newBuilder()
        .uri(URI.create(rpcUrl))
        .header("Content-Type", "application/json")
        .POST(BodyPublishers.ofString(request.toJson.toString))
        .build()

      val response = client.send(httpRequest, BodyHandlers.ofString())

      if (response.statusCode() != 200) {
        throw new Exception(s"HTTP error: ${response.statusCode()}")
      }

      val json = response.body().parseJson.asJsObject
      json.fields.get("result") match {
        case Some(result) => result.convertTo[T]
        case None =>
          val error = json.fields.get("error")
          throw new Exception(s"RPC error: $error")
      }
    }

    /**
     * Get latest block number
     */
    def getBlockNumber: Try[Long] = {
      rpcCall[String]("eth_blockNumber")
        .map(hex => java.lang.Long.parseLong(hex.stripPrefix("0x"), 16))
    }

    /**
     * Get balance of address in ETH
     */
    def getBalance(address: String): Try[BigDecimal] = {
      rpcCall[String]("eth_getBalance", List(address, "latest"))
        .map { hex =>
          val wei = BigInt(hex.stripPrefix("0x"), 16)
          BigDecimal(wei) / BigDecimal("1000000000000000000")
        }
    }

    /**
     * Get current gas price in Gwei
     */
    def getGasPrice: Try[BigDecimal] = {
      rpcCall[String]("eth_gasPrice")
        .map { hex =>
          val wei = BigInt(hex.stripPrefix("0x"), 16)
          BigDecimal(wei) / BigDecimal("1000000000")
        }
    }

    /**
     * Get transaction by hash
     */
    def getTransaction(txHash: String): Try[JsValue] = {
      rpcCall[JsValue]("eth_getTransactionByHash", List(txHash))
    }

    /**
     * Get block by number
     */
    def getBlock(blockNumber: Long): Try[JsValue] = {
      val blockHex = s"0x${blockNumber.toHexString}"
      rpcCall[JsValue]("eth_getBlockByNumber", List(blockHex, true))
    }

    /**
     * Call contract method (read-only)
     */
    def call(to: String, data: String, from: Option[String] = None): Try[String] = {
      val params = Map(
        "to" -> to.toJson,
        "data" -> data.toJson
      ) ++ from.map(f => "from" -> f.toJson)

      rpcCall[String]("eth_call", List(params.toJson, "latest"))
    }
  }

  def main(args: Array[String]): Unit = {
    println("⚙️  Scala Web3 Client")
    println("-" * 40)

    val client = new EthereumClient()

    // Get block number
    client.getBlockNumber match {
      case Success(block) => println(s"📦 Block Number: ${block}")
      case Failure(e) => println(s"❌ Error: ${e.getMessage}")
    }

    // Get gas price
    client.getGasPrice match {
      case Success(gas) => println(f"⛽ Gas Price: ${gas}%.2f Gwei")
      case Failure(e) => println(s"❌ Error: ${e.getMessage}")
    }

    // Get balance (Vitalik's address)
    val address = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"
    client.getBalance(address) match {
      case Success(balance) => println(f"💰 Balance: ${balance}%.4f ETH")
      case Failure(e) => println(s"❌ Error: ${e.getMessage}")
    }
  }
}

File: examples/fsharp/defi-analytics/Web3Client.fs
// F# Web3 Client - Functional blockchain analytics
module Web3Client

open System
open System.Net.Http
open System.Text
open System.Text.Json
open System.Numerics

type RpcRequest = {
    jsonrpc: string
    method: string
    params: obj[]
    id: int
}

type RpcResponse<'T> = {
    jsonrpc: string
    result: 'T option
    error: JsonElement option
    id: int
}

type Web3Client(rpcUrl: string) =
    let mutable requestId = 0
    let httpClient = new HttpClient()

    member private _.RpcCall<'T>(method: string, [<ParamArray>] params: obj[]) : Async<'T> =
        async {
            requestId <- requestId + 1

            let payload = {
                jsonrpc = "2.0"
                method = method
                params = params
                id = requestId
            }

            let json = JsonSerializer.Serialize(payload)
            let content = new StringContent(json, Encoding.UTF8, "application/json")

            let! response = httpClient.PostAsync(rpcUrl, content) |> Async.AwaitTask
            let! body = response.Content.ReadAsStringAsync() |> Async.AwaitTask

            if not response.IsSuccessStatusCode then
                failwithf "HTTP error: %d" (int response.StatusCode)

            let result = JsonSerializer.Deserialize<RpcResponse<'T>>(body)

            match result.result with
            | Some r -> return r
            | None -> return failwith "RPC error"
        }

    member private _.HexToInt(hex: string) : bigint =
        BigInteger.Parse(hex.Replace("0x", ""), Globalization.NumberStyles.HexNumber)

    member private _.WeiToEther(wei: bigint) : decimal =
        decimal wei / 1_000_000_000_000_000_000m

    member private _.WeiToGwei(wei: bigint) : decimal =
        decimal wei / 1_000_000_000m

    member this.GetBlockNumberAsync() : Async<bigint> =
        async {
            let! result = this.RpcCall<string>("eth_blockNumber")
            return this.HexToInt(result)
        }

    member this.GetBalanceAsync(address: string) : Async<decimal> =
        async {
            let! result = this.RpcCall<string>("eth_getBalance", address, "latest")
            let wei = this.HexToInt(result)
            return this.WeiToEther(wei)
        }

    member this.GetGasPriceAsync() : Async<decimal> =
        async {
            let! result = this.RpcCall<string>("eth_gasPrice")
            let wei = this.HexToInt(result)
            return this.WeiToGwei(wei)
        }

    member this.GetTransactionAsync(txHash: string) : Async<JsonElement> =
        this.RpcCall<JsonElement>("eth_getTransactionByHash", txHash)

[<EntryPoint>]
let main argv =
    printfn "🔷 F# Web3 Client"
    printfn "%s" (String.replicate 40 "-")

    let client = Web3Client("https://eth.llamarpc.com")

    try
        // Get block number
        let block = client.GetBlockNumberAsync() |> Async.RunSynchronously
        printfn "📦 Block Number: %A" block

        // Get gas price
        let gas = client.GetGasPriceAsync() |> Async.RunSynchronously
        printfn "⛽ Gas Price: %.2f Gwei" gas

        // Get balance
        let address = "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045"
        let balance = client.GetBalanceAsync(address) |> Async.RunSynchronously
        printfn "💰 Balance: %.4f ETH" balance

        0
    with
    | ex ->
        printfn "❌ Error: %s" ex.Message
        1
